src/interchange/src/avro.rs

```diff
@@ -798,7 +799,33 @@ impl<'a> AvroDecode for AvroFlatDecoder<'a> {
         *self.buf = str_buf;
         Ok(())
     }
-    define_unexpected! {map}
+    #[inline]
+    fn map<A: AvroMapAccess>(self, a: &mut A) -> Result<Self::Out, AvroError> {
+        let mut map_buf = std::mem::take(self.buf);
+        // Map (key, value) pairs need to be unique and ordered.
+        let mut map = BTreeMap::new();
+        while let Some((name, f)) = a.next_entry()? {
+            map.insert(name, f.decode_field(ValueDecoder)?);
+        }
+        self.packer
+            .push_dict_with(|packer| -> Result<(), AvroError> {
+                for (key, val) in map {
+                    packer.push(Datum::String(key.as_str()));
+                    give_value(
+                        AvroFlatDecoder {
+                            packer,
+                            buf: &mut map_buf,
```

Comment:
`buf` will never actually be used, since you are passing in already-decoded values. So you can just give it `&mut vec![]` here with no negative performance effects, and avoid taking and resetting the buffer on lines 804 and 826.